source/common/stats/stats_impl.cc

```diff
@@ -135,35 +135,47 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
   return false;
 }
 
-RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
-  uint64_t num_bytes_to_allocate = RawStatData::structSize(name.size());
-  RawStatData* data = static_cast<RawStatData*>(::calloc(num_bytes_to_allocate, 1));
-  if (data == nullptr) {
-    throw std::bad_alloc();
-  }
-  data->checkAndInit(name, num_bytes_to_allocate);
+HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
 
+HeapStatData* HeapStatDataAllocator::alloc(const std::string& name) {
+  auto data = std::make_unique<HeapStatData>(truncateStatName(name));
```

Comment:
I didn't see this behavior change documented in the PR description - are we now going back to truncating in non hot restart mode?  I think we should probably think carefully about what we want the user-facing behavior to look like long-term before changing the behavior again to avoid churn for something that is at least nominally user-facing.  As @mattklein123 has noted in the past, this behavior has become quite complex, so I think it's worth some thought to simplify the user-facing surface and document the behavior.

Google internally and (I assume) most other users of Envoy in non hot restart mode have no reason to want stats to be truncated.  I think it's totally okay to allow truncation in that mode, but if we do, I think it would be better to allow it to be configurable via the command line.  We could try to incorporate this optionality into the existing command line parameter somehow (for instance, allow users to pass in 0 for no truncation).  WDYT?

Side note: I'm less concerned about the isolated stats store as those stats are not exported to or queryable by the user unless made available through other means.  We could add a way to pass in stats options from whatever code is creating it, but I see no reason why it needs to conform to the parameters used for the primary stats store as it's meant to be a distinct entity.

Reply:
I think this is actually not a behavior change; we are truncating prior to this PR in thread_local_store, but not IsolatedStore.  It follows the configurable options.

Reply:
https://github.com/envoyproxy/envoy/issues/3965

Reply:
I'm not seeing where we're doing the truncation for the non-hot restart case.  Can you please point me to it?  [Here](https://github.com/envoyproxy/envoy/blob/324e628b7c85bcbeacfb49050dd7bfdcc86702b4/source/common/stats/thread_local_store.cc#L181) we seem to be passing through the full name, and I can't seem to find anywhere that we actually perform truncation (unless we're falling back in the hot restart case) in thread_local_store.

Reply:
@mrice32 : HeapStatDataAllocator::alloc() truncates in the first line of its impl:
```
HeapStatData* HeapStatDataAllocator::alloc(const std::string& name) {
  auto data = std::make_unique<HeapStatData>(truncateStatName(name));
  ...
```

Reply:
>I think this is actually not a behavior change; we are truncating prior to this PR in thread_local_store, but not IsolatedStore. It follows the configurable options.

I was referring to the code before your change.  I think we were not truncating before, and you changed the behavior such that we are.  I'm looking for where we were truncating in the non hot restart case before your change, and I can't find it.